python/pyarrow/tests/test_io.py

```diff
@@ -160,6 +188,24 @@ def test_inmemory_write_after_closed():
     with pytest.raises(IOError):
         f.write(b'not ok')
 
+def test_buffer_protocol_ref_counting():
+    import gc
+
+    val = b'data'
+    buf = io.buffer_from_bytes(val)
+    as_bytes = bytes(buf)
+
+    def add_temporary_reference(buf):
+        m = bytearray(buf)
+        del(buf)
```

Comment:
Gotcha, the thing I worry about with these things is a broken parent-child relationship